### PR TITLE
[perf] Disable inline_in_all_cgus when LTO is enabled

### PR DESCRIPTION
**Pull request by ishitatsuyuki**
I tried to run lolbench with this flag altered; however lolbench either errored or got stuck, so I end up running only 3 of the benchmarks manually. One showed no changes, another one showed 20% perf improvement somehow, and another showed slight (1%) regression.

So I'm thinking of benchmarking it with perf instead; if we still have ThinLTO on CI then we can measure stage2 compiler performance as a metric of how much it affects generated code.

### CODE REVIEW

**Comment by ishitatsuyuki** [comment]
cc @Mark-Simulacrum

**Comment by Mark-Simulacrum** [comment]
@bors try @rust-timer queue

A perf run can't hurt, I guess.

**Comment by ishitatsuyuki** [comment]
Everything within the margin of noise, is ThinLTO used for benchmarks?

**Comment by Mark-Simulacrum** [comment]
It should be, yes, for release builds by default I believe.

**Comment by bjorn3** [comment]
@ishitatsuyuki `@rust-timer` only benchmarks compilation time, so this means that rustc perf didn't change. lolbench is a different program, which does benchmark execution time.

**Comment by ishitatsuyuki** [comment]
@bjorn3 I am aware of that, and what I expected is that this change will indirectly affect stage2 compiler, which in turn can be benchmarked for performance difference.

So far what I have heard is that ThinLTO is enabled for both compiler builds and perf, so everything being identical sounds really strange. Something more likely is that the code/flag being changed has no effect at all, so maybe I should ask who had implemented this originally -

r? @alexcrichton

File: src/librustc_mir/monomorphize/item.rs [hunk 1]
```diff
@@ -65,7 +65,7 @@ pub trait MonoItemExt<'tcx>: fmt::Debug {
     fn instantiation_mode(&self, tcx: TyCtxt<'tcx>) -> InstantiationMode {
         let inline_in_all_cgus =
             tcx.sess.opts.debugging_opts.inline_in_all_cgus.unwrap_or_else(|| {
-                tcx.sess.opts.optimize != OptLevel::No
+                tcx.sess.opts.optimize != OptLevel::No && if tcx.sess.lto() == Lto::No
```
Comment (nikic) [inline src/librustc_mir/monomorphize/item.rs]: There is an `if` after the `&&`. Why does this code compile?

Comment (ishitatsuyuki) [reply src/librustc_mir/monomorphize/item.rs]: Oh good catch, and yeah I have no idea why this happened to compile. I'll push a fix later.

Comment (bjorn3) [reply src/librustc_mir/monomorphize/item.rs]: Strange. I can't reproduce on play.rust-lang.org.

**Comment by ishitatsuyuki** [comment]
Oh shit this file is dead code. The code is inside `src/librustc/mir/mono.rs` instead.

**Comment by ishitatsuyuki** [comment]
@Mark-Simulacrum Sorry, can I get another perf run?

**Comment by ishitatsuyuki** [comment]
Checked this and it seems it’s mostly a big lose. Closing.

**Pull request closed without merging**